src/formula/variant.cpp

```diff
@@ -941,141 +686,54 @@ variant variant::concatenate(const variant& v) const
 	}
 
-	if(type_ == TYPE_LIST) {
-		variant_iterator iter = std::find(begin(), end(), v);
-		return iter != end();
+	if(type() == VARIANT_TYPE::TYPE_LIST) {
+		return value_cast<game_logic::variant_list>()->contains(v);
 	} else {
-		std::map<variant,variant>::const_iterator iter = map_->find(v);
-		return iter != map_->end();
+		return value_cast<game_logic::variant_map>()->contains(v);
 	}
 }
 
-void variant::must_be(variant::TYPE t) const
+void variant::must_be(VARIANT_TYPE t) const
 {
-	if(type_ != t) {
-		throw type_error(formatter() << "type error: " << " expected "
-			<< variant_type_to_string(t) << " but found " << type_string()
-			<< " (" << to_debug_string() << ")");
+	if(type() != t) {
+		throw type_error(formatter() << "type error: expected "
+			<< variant_type_to_string(t) << " but found "
+			<< type_string() << " (" << to_debug_string() << ")");
 	}
 }
 
-void variant::serialize_to_string(std::string& str) const
+void variant::must_both_be(VARIANT_TYPE t, const variant& second) const
 {
-	switch(type_) {
-	case TYPE_NULL:
-		str += "null()";
-		break;
-	case TYPE_INT:
-		str += std::to_string(int_value_);
-		break;
-	case TYPE_DECIMAL: {
-		std::ostringstream s;
-
-		int fractional = decimal_value_ % 1000;
-		int integer = (decimal_value_ - fractional) / 1000;
-
-		s << integer << ".";
-
-		fractional = std::abs(fractional);
-
-		if( fractional < 100) {
-			if( fractional < 10)
-				s << "00";
-			else
-				s << 0;
-		}
-
-		s << fractional;
-
-		str += s.str();
-		break;
-	}
-	case TYPE_CALLABLE:
-		callable_->serialize(str);
-		break;
-	case TYPE_LIST: {
-		str += "[";
-		bool first_time = true;
-		for(size_t i=0; i<list_->size(); ++i) {
-			const variant& var = (*list_)[i];
-			if(!first_time) {
-				str += ",";
-			}
-			first_time = false;
-			var.serialize_to_string(str);
-		}
-		str += "]";
-		break;
-	}
-	case TYPE_MAP: {
-		str += "[";
-		bool first_time = true;
-		for(std::map<variant,variant>::const_iterator i=map_->begin(); i != map_->end(); ++i) {
-			if(!first_time) {
-				str += ",";
-			}
-			first_time = false;
-			i->first.serialize_to_string(str);
-			str += "->";
-			i->second.serialize_to_string(str);
-		}
-		if(map_->empty()) {
-			str += "->";
-		}
-		str += "]";
-		break;
-	}
-	case TYPE_STRING:
-		str += "'";
-		for(std::string::iterator it = string_->begin(); it < string_->end(); ++it) {
-			switch(*it) {
-			case '\'':
-				str += "[']";
-				break;
-			case '[':
-				str += "[(]";
-				break;
-			case ']':
-				str += "[)]";
-				break;
-			default:
-				str += *it;
-				break;
-			}
-		}
-		str += "'";
-		break;
-	default:
-		assert(false);
+	if(type() != t || second.type() != t) {
+		throw type_error(formatter() << "type error: expected "
+			<< variant_type_to_string(t) << " but found "
+			<<        type_string() << " (" <<        to_debug_string() << ")" << " and "
+			<< second.type_string() << " (" << second.to_debug_string() << ")");
 	}
 }
 
+std::string variant::serialize_to_string() const
```

Comment:
Again... why did you change this API to return the string instead of taking it as a reference?